File: src/ipc/provider_eligibility.ts
import { isRecord } from "./errors";
import type { ProviderEligibilityDto } from "./types";

const WORK_KINDS = ["planning", "implementation", "review"] as const;
const PROVIDERS = ["claude", "codex"] as const;
const CAPABILITIES = ["supported", "unsupported", "unavailable"] as const;
const CONTRACTS = ["approved", "notApproved"] as const;
const BLOCKING_REASONS = [
  "capabilityUnavailable",
  "capabilityUnsupported",
  "contractNotApproved",
  "taskStateMismatch",
] as const;
const ENTRY_KEYS = [
  "workKind",
  "provider",
  "capability",
  "contract",
  "eligible",
  "stateAllowsWorkKind",
  "blockingReasons",
] as const;

export function isProviderEligibilityDtoArray(
  value: unknown,
): value is readonly ProviderEligibilityDto[] {
  if (
    !Array.isArray(value) ||
    value.length !== WORK_KINDS.length * PROVIDERS.length
  ) {
    return false;
  }
  const combinations = new Set<string>();
  for (const entry of value) {
    if (!isProviderEligibilityDto(entry)) {
      return false;
    }
    combinations.add(`${entry.workKind}:${entry.provider}`);
  }
  return combinations.size === WORK_KINDS.length * PROVIDERS.length;
}

function isProviderEligibilityDto(value: unknown): value is ProviderEligibilityDto {
  if (!isRecord(value)) {
    return false;
  }
  const keys = Object.keys(value);
  return (
    keys.length === ENTRY_KEYS.length &&
    keys.every((key) => ENTRY_KEYS.some((allowed) => allowed === key)) &&
    isAllowedString(value.workKind, WORK_KINDS) &&
    isAllowedString(value.provider, PROVIDERS) &&
    isAllowedString(value.capability, CAPABILITIES) &&
    isAllowedString(value.contract, CONTRACTS) &&
    typeof value.eligible === "boolean" &&
    typeof value.stateAllowsWorkKind === "boolean" &&
    Array.isArray(value.blockingReasons) &&
    value.blockingReasons.every((reason) => isAllowedString(reason, BLOCKING_REASONS))
  );
}

function isAllowedString(value: unknown, allowed: readonly string[]): boolean {
  return typeof value === "string" && allowed.includes(value);
}

File: crates/chatoms-application/src/provider_eligibility.rs
use chatoms_domain::{TaskState, WorkKind};
use chatoms_ports::provider::ProviderKind;

use crate::system::{CapabilityStatus, ProviderCapabilitySummary};

const PROVIDERS: [ProviderKind; 2] = [ProviderKind::Claude, ProviderKind::Codex];

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum EligibilityCapability {
    Supported,
    Unsupported,
    Unavailable,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ContractStatus {
    Approved,
    NotApproved,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum EligibilityBlockingReason {
    CapabilityUnavailable,
    CapabilityUnsupported,
    ContractNotApproved,
    TaskStateMismatch,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ProviderEligibilityView {
    pub work_kind: WorkKind,
    pub provider: ProviderKind,
    pub capability: EligibilityCapability,
    pub contract: ContractStatus,
    pub eligible: bool,
    pub state_allows_work_kind: bool,
    pub blocking_reasons: Vec<EligibilityBlockingReason>,
}

pub struct ProviderEligibilityPolicy;

impl ProviderEligibilityPolicy {
    #[must_use]
    pub fn evaluate(
        task_state: TaskState,
        capabilities: ProviderCapabilitySummary,
    ) -> Vec<ProviderEligibilityView> {
        let mut entries = Vec::with_capacity(WorkKind::ALL.len() * PROVIDERS.len());
        for work_kind in WorkKind::ALL {
            for provider in PROVIDERS {
                let capability = capability_for(capabilities, provider);
                let contract = contract_for(provider, work_kind);
                let eligible = capability == EligibilityCapability::Supported
                    && contract == ContractStatus::Approved;
                let state_allows_work_kind = work_kind.can_start_from(task_state);
                entries.push(ProviderEligibilityView {
                    work_kind,
                    provider,
                    capability,
                    contract,
                    eligible,
                    state_allows_work_kind,
                    blocking_reasons: blocking_reasons(
                        capability,
                        contract,
                        state_allows_work_kind,
                    ),
                });
            }
        }
        entries
    }
}

const fn capability_for(
    capabilities: ProviderCapabilitySummary,
    provider: ProviderKind,
) -> EligibilityCapability {
    let capability = match provider {
        ProviderKind::Claude => capabilities.claude,
        ProviderKind::Codex => capabilities.codex,
    };
    match capability {
        Some(CapabilityStatus::Supported) => EligibilityCapability::Supported,
        Some(CapabilityStatus::Unsupported) => EligibilityCapability::Unsupported,
        None => EligibilityCapability::Unavailable,
    }
}

const fn contract_for(provider: ProviderKind, work_kind: WorkKind) -> ContractStatus {
    match (provider, work_kind) {
        (
            ProviderKind::Claude,
            WorkKind::Planning | WorkKind::Implementation | WorkKind::Review,
        ) => ContractStatus::Approved,
        (ProviderKind::Codex, _) => ContractStatus::NotApproved,
    }
}

fn blocking_reasons(
    capability: EligibilityCapability,
    contract: ContractStatus,
    state_allows_work_kind: bool,
) -> Vec<EligibilityBlockingReason> {
    let mut reasons = Vec::with_capacity(3);
    match capability {
        EligibilityCapability::Supported => {}
        EligibilityCapability::Unsupported => {
            reasons.push(EligibilityBlockingReason::CapabilityUnsupported);
        }
        EligibilityCapability::Unavailable => {
            reasons.push(EligibilityBlockingReason::CapabilityUnavailable);
        }
    }
    match contract {
        ContractStatus::Approved => {}
        ContractStatus::NotApproved => {
            reasons.push(EligibilityBlockingReason::ContractNotApproved);
        }
    }
    if !state_allows_work_kind {
        reasons.push(EligibilityBlockingReason::TaskStateMismatch);
    }
    reasons
}
